src/prover.rs

```diff
@@ -352,99 +387,68 @@ impl<LeafStorage: LeafCache> Prover<LeafStorage> {
                 self.leaf_data.cache_size(),
                 block.txdata.len()
             );
-
-            let (proof, leaves) = self.process_block(&block, height);
-            let block = bitcoin::network::utreexo::UtreexoBlock {
-                block,
-                udata: Some(UData {
-                    remember_idx: vec![],
-                    proof,
-                    leaves,
-                }),
-            };
-            let index = self.files.write().unwrap().append(&block);
-            self.storage.append(index, block.block.block_hash());
+            let mtp = self.rpc.get_mtp(block.block_hash())?;
+            let (proof, leaves) = self.process_block(&block, height, mtp);
+            let index = self
+                .files
+                .write()
+                .unwrap()
+                .save_block(&block, height, proof, leaves, &self.acc);
+
+            self.storage.append(index, block.block_hash());
             self.height = height;
+            if let Some(n) = self.snapshot_acc_every {
+                if height % n == 0 {
+                    self.save_to_disk(Some(height))
+                        .expect("could not save the acc to disk");
+                }
+            }
         }
         anyhow::Ok(())
     }
 
-    /// Returns what spk type this output is. We need this to build a compact leaf data.
-    fn get_spk_type(spk: &Script) -> ScriptPubkeyType {
-        if spk.is_p2pkh() {
-            ScriptPubkeyType::PubKeyHash
-        } else if spk.is_p2sh() {
-            ScriptPubkeyType::ScriptHash
-        } else if spk.is_v0_p2wpkh() {
-            ScriptPubkeyType::WitnessV0PubKeyHash
-        } else if spk.is_v0_p2wsh() {
-            ScriptPubkeyType::WitnessV0ScriptHash
-        } else {
-            ScriptPubkeyType::Other(spk.to_bytes().into_boxed_slice())
-        }
-    }
-
     /// Pulls the [LeafData] from the bitcoin core rpc. We use this as fallback if we can't find
     /// the leaf in leaf_data. This method is slow and should only be used if we can't find the
     /// leaf in the leaf_data.
-    fn get_input_leaf_hash_from_rpc(rpc: &Box<dyn Blockchain>, input: &TxIn) -> Option<LeafData> {
+    fn get_input_leaf_hash_from_rpc(rpc: &dyn Blockchain, input: &TxIn) -> Option<LeafContext> {
         let tx_info = rpc
             .get_raw_transaction_info(&input.previous_output.txid)
             .ok()?;
+
         let height = tx_info.height;
         let output = &tx_info.tx.output[input.previous_output.vout as usize];
-        let header_code = if tx_info.is_coinbase {
-            height << 1 | 1
-        } else {
-            height << 1
-        };
-        Some(LeafData {
+        let median_time_past = rpc.get_mtp(tx_info.blockhash?).ok()?;
```

Comment:
Here it should be the MTP of the block prior to the one including the tx: https://github.com/bitcoin/bitcoin/blob/712a2b5453cdf2568fece94b969d6e0923b6ba16/src/consensus/tx_verify.cpp#L87